File: internal/analyzers/common/threshold_labeler.go
package common

// ThresholdLabeler maps a float64 score to a string label using an ordered list
// of Threshold[float64] values. Thresholds must be sorted descending by Limit
// (highest first) — the first threshold where score >= Limit wins.
// A catch-all fallback can be added as the last entry with Limit set to 0 or
// the minimum possible score.
//
// Example:
//
//	labeler := common.ThresholdLabeler{
//	    {Limit: 0.8, Label: "Excellent"},
//	    {Limit: 0.6, Label: "Good"},
//	    {Limit: 0.4, Label: "Fair"},
//	    {Limit: 0.0, Label: "Poor"},
//	}
//	labeler.Label(0.75) // → "Good"
type ThresholdLabeler []Threshold[float64]

// Label returns the label of the first threshold where score >= Limit.
// Returns "" if the labeler is empty or no threshold matches.
func (l ThresholdLabeler) Label(score float64) string {
	for _, t := range l {
		if score >= t.Limit {
			return t.Label
		}
	}

	return ""
}

File: rust/crates/cf-analyzers-common/src/threshold_labeler.rs
//! Float-score-to-label mapping (`threshold_labeler.go`).
//!
//! A thin wrapper over an ordered list of [`Threshold<f64>`] values. Unlike
//! [`crate::classify::Classifier`], the labeler does **not** sort its
//! thresholds: callers must supply them in descending order by limit, and the
//! first threshold where `score >= limit` wins. An empty labeler, or a score
//! below every limit, yields the empty string.

use crate::classify::Threshold;

/// Maps a `f64` score to a string label using a caller-ordered threshold list.
///
/// Mirrors `common.ThresholdLabeler`. Thresholds must be sorted descending by
/// limit (highest first); a catch-all can be added as a final entry with the
/// minimum limit.
#[derive(Debug, Clone, Default)]
pub struct ThresholdLabeler {
    thresholds: Vec<Threshold<f64>>,
}

impl ThresholdLabeler {
    /// Creates a labeler from an already-descending threshold list.
    #[must_use]
    pub fn new(thresholds: Vec<Threshold<f64>>) -> Self {
        ThresholdLabeler { thresholds }
    }

    /// Returns the label of the first threshold where `score >= limit`.
    ///
    /// Returns `""` if the labeler is empty or no threshold matches. Mirrors
    /// `common.ThresholdLabeler.Label`.
    #[must_use]
    pub fn label(&self, score: f64) -> &str {
        for t in &self.thresholds {
            if score >= t.limit {
                return &t.label;
            }
        }
        ""
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn labeler() -> ThresholdLabeler {
        ThresholdLabeler::new(vec![
            Threshold::new(0.8, "Excellent"),
            Threshold::new(0.6, "Good"),
            Threshold::new(0.4, "Fair"),
            Threshold::new(0.0, "Poor"),
        ])
    }

    #[test]
    fn label_boundaries() {
        let l = labeler();
        for (score, want) in [
            (0.9, "Excellent"),
            (0.8, "Excellent"),
            (0.7, "Good"),
            (0.6, "Good"),
            (0.5, "Fair"),
            (0.4, "Fair"),
            (0.1, "Poor"),
            (0.0, "Poor"),
        ] {
            assert_eq!(l.label(score), want, "score {score}");
        }
    }

    #[test]
    fn empty_labeler_returns_empty() {
        let l = ThresholdLabeler::default();
        assert_eq!(l.label(0.5), "");
    }

    #[test]
    fn no_match_returns_empty() {
        let l = ThresholdLabeler::new(vec![Threshold::new(0.8, "high")]);
        assert_eq!(l.label(0.5), "");
    }
}
